<a href="https://colab.research.google.com/github/boufousabd/leaf-classification/blob/main/notebooks/11182020-leaf-classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
leaf_train_df = pd.read_csv("../data/raw/train.csv",index_col='id')
leaf_test_df = pd.read_csv("../data/raw/test.csv",index_col='id')

In [ ]:
leaf_train_df

species   margin1   margin2   margin3   margin4   margin5  \
id                                                                              
1               Acer_Opalus  0.007812  0.023438  0.023438  0.003906  0.011719   
2     Pterocarya_Stenoptera  0.005859  0.000000  0.031250  0.015625  0.025391   
3      Quercus_Hartwissiana  0.005859  0.009766  0.019531  0.007812  0.003906   
5           Tilia_Tomentosa  0.000000  0.003906  0.023438  0.005859  0.021484   
6        Quercus_Variabilis  0.005859  0.003906  0.048828  0.009766  0.013672   
...                     ...       ...       ...       ...       ...       ...   
1575   Magnolia_Salicifolia  0.060547  0.119140  0.007812  0.003906  0.000000   
1578            Acer_Pictum  0.001953  0.003906  0.021484  0.107420  0.001953   
1581     Alnus_Maximowiczii  0.001953  0.003906  0.000000  0.021484  0.078125   
1582          Quercus_Rubra  0.000000  0.000000  0.046875  0.056641  0.009766   
1584         Quercus_Afares  0.023438  0.019531  0.031250  0.015625  0.005859   

       margin6   margin7  margin8   margin9  ...  texture55  texture56  \
id                                           ...                         
1     0.009766  0.027344      0.0  0.001953  ...   0.007812   0.000000   
2     0.001953  0.019531      0.0  0.000000  ...   0.000977   0.000000   
3     0.005859  0.068359      0.0  0.000000  ...   0.154300   0.000000   
5     0.019531  0.023438      0.0  0.013672  ...   0.000000   0.000977   
6     0.015625  0.005859      0.0  0.000000  ...   0.096680   0.000000   
...        ...       ...      ...       ...  ...        ...        ...   
1575  0.148440  0.017578      0.0  0.001953  ...   0.242190   0.000000   
1578  0.000000  0.000000      0.0  0.029297  ...   0.170900   0.000000   
1581  0.003906  0.007812      0.0  0.003906  ...   0.004883   0.000977   
1582  0.000000  0.000000      0.0  0.037109  ...   0.083008   0.030273   
1584  0.019531  0.035156      0.0  0.003906  ...   0.000000   0.000000   

      texture57  texture58  texture59  texture60  texture61  texture62  \
id                                                                       
1      0.002930   0.002930   0.035156   0.000000   0.000000   0.004883   
2      0.000000   0.000977   0.023438   0.000000   0.000000   0.000977   
3      0.005859   0.000977   0.007812   0.000000   0.000000   0.000000   
5      0.000000   0.000000   0.020508   0.000000   0.000000   0.017578   
6      0.021484   0.000000   0.000000   0.000000   0.000000   0.000000   
...         ...        ...        ...        ...        ...        ...   
1575   0.034180   0.000000   0.010742   0.000000   0.000000   0.000000   
1578   0.018555   0.000000   0.011719   0.000000   0.000000   0.000977   
1581   0.004883   0.027344   0.016602   0.007812   0.000000   0.027344   
1582   0.000977   0.002930   0.014648   0.000000   0.041992   0.000000   
1584   0.002930   0.000000   0.012695   0.000000   0.000000   0.023438   

      texture63  texture64  
id                          
1      0.000000   0.025391  
2      0.039062   0.022461  
3      0.020508   0.002930  
5      0.000000   0.047852  
6      0.000000   0.031250  
...         ...        ...  
1575   0.000000   0.018555  
1578   0.000000   0.021484  
1581   0.000000   0.001953  
1582   0.001953   0.002930  
1584   0.025391   0.022461  

[990 rows x 193 columns]

In [ ]:
leaf_test_df

margin1   margin2   margin3   margin4   margin5   margin6   margin7  \
id                                                                           
4     0.019531  0.009766  0.078125  0.011719  0.003906  0.015625  0.005859   
7     0.007812  0.005859  0.064453  0.009766  0.003906  0.013672  0.007812   
9     0.000000  0.000000  0.001953  0.021484  0.041016  0.000000  0.023438   
12    0.000000  0.000000  0.009766  0.011719  0.017578  0.000000  0.003906   
13    0.001953  0.000000  0.015625  0.009766  0.039062  0.000000  0.009766   
...        ...       ...       ...       ...       ...       ...       ...   
1576  0.000000  0.000000  0.003906  0.015625  0.041016  0.000000  0.017578   
1577  0.000000  0.003906  0.003906  0.005859  0.017578  0.000000  0.017578   
1579  0.017578  0.029297  0.015625  0.013672  0.003906  0.015625  0.025391   
1580  0.013672  0.009766  0.060547  0.025391  0.035156  0.025391  0.039062   
1583  0.000000  0.117190  0.000000  0.019531  0.000000  0.136720  0.001953   

       margin8   margin9  margin10  ...  texture55  texture56  texture57  \
id                                  ...                                    
4     0.000000  0.005859  0.023438  ...   0.006836   0.000000   0.015625   
7     0.000000  0.033203  0.023438  ...   0.000000   0.000000   0.006836   
9     0.000000  0.011719  0.005859  ...   0.128910   0.000000   0.000977   
12    0.000000  0.003906  0.001953  ...   0.012695   0.015625   0.002930   
13    0.000000  0.005859  0.000000  ...   0.000000   0.042969   0.016602   
...        ...       ...       ...  ...        ...        ...        ...   
1576  0.000000  0.005859  0.013672  ...   0.098633   0.000000   0.004883   
1577  0.005859  0.000000  0.005859  ...   0.012695   0.004883   0.004883   
1579  0.000000  0.000000  0.009766  ...   0.073242   0.000000   0.028320   
1580  0.000000  0.003906  0.023438  ...   0.003906   0.000000   0.000977   
1583  0.005859  0.000000  0.007812  ...   0.107420   0.012695   0.016602   

      texture58  texture59  texture60  texture61  texture62  texture63  \
id                                                                       
4      0.000977   0.015625        0.0        0.0   0.000000   0.003906   
7      0.001953   0.013672        0.0        0.0   0.000977   0.037109   
9      0.000000   0.000000        0.0        0.0   0.015625   0.000000   
12     0.036133   0.013672        0.0        0.0   0.089844   0.000000   
13     0.010742   0.041016        0.0        0.0   0.007812   0.009766   
...         ...        ...        ...        ...        ...        ...   
1576   0.000000   0.003906        0.0        0.0   0.018555   0.000000   
1577   0.002930   0.009766        0.0        0.0   0.090820   0.000000   
1579   0.000000   0.001953        0.0        0.0   0.000000   0.042969   
1580   0.000000   0.011719        0.0        0.0   0.000000   0.011719   
1583   0.000977   0.004883        0.0        0.0   0.015625   0.000000   

      texture64  
id               
4      0.053711  
7      0.044922  
9      0.000000  
12     0.008789  
13     0.007812  
...         ...  
1576   0.000977  
1577   0.016602  
1579   0.006836  
1580   0.018555  
1583   0.017578  

[594 rows x 192 columns]

In [ ]:
train_data, train_target = leaf_train_df[leaf_train_df.columns.difference(['species'])], leaf_train_df['species']

## 192 dimensions, 99 classes, 990 données train, 594 données de test (1,584 en total)

## SVM Linear

In [ ]:
from sklearn import svm
from sklearn.model_selection import GridSearchCV

In [ ]:
C = [1, 10, 100, 1000]

In [ ]:
parameters_grid = [
    {'C': [1, 10, 100, 1000], 'kernel': ['linear']},
    {'C': [1, 10, 100, 1000], 'kernel': ['poly'], 'degree':list(range(2,9)), 'coef0':list(np.linspace(0.000001, 2, num=10))},
    {'C': [1, 10, 100, 1000], 'kernel': ['rbf'], 'gamma':['scale','auto'], 'gamma':list(np.linspace(0.000001, 2, num=10))},
    {'C': [1, 10, 100, 1000], 'kernel': ['sigmoid'], 'coef0':list(np.linspace(0.000001, 2, num=10))}
]

In [ ]:
svc = svm.SVC()
clf = GridSearchCV(svc, parameters_grid, cv=10, n_jobs=-1)
clf.fit(train_data, train_target)

In [ ]:
best_params = clf.best_params_
best_params

In [ ]:
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score

In [ ]:
clf_linear = svm.SVC(kernel= 'linear', C = 100)
scores_linear = cross_val_score(clf_linear, train_data, train_target, cv=10, scoring='f1_macro')
print("F1_Macro Linear",scores_linear.mean()," (+/-) ", scores_linear.std() * 2)

F1_Macro Linear 0.9437710437710436  (+/-)  0.051641534195388084


In [ ]:
clf_rbf = svm.SVC(kernel= 'rbf', gamma = 0.4444452222222222, C = 100)
scores_rbf = cross_val_score(clf_rbf, train_data, train_target, cv=10, scoring='f1_macro')
print("F1_Macro Rbf",scores_rbf.mean()," (+/-) ", scores_rbf.std() * 2)

F1_Macro Rbf 0.9447811447811446  (+/-)  0.050320651022947846


## Gradient Descent

In [ ]:
#alpha, optimal,
from sklearn.linear_model import SGDClassifier
tuned_parameters = [{'alpha': list(np.linspace(0.000001, 2, num = 10))}]
clf_sgd = SGDClassifier(max_iter=10000, penalty='l1', shuffle=True, loss='hinge', tol=1e-3)
clf_sgd_grid_search = make_pipeline(StandardScaler(), GridSearchCV(clf_sgd, tuned_parameters, scoring = 'f1_macro', cv=10))
clf_sgd_grid_search.fit(train_data, train_target)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('gridsearchcv',
                 GridSearchCV(cv=10,
                              estimator=SGDClassifier(max_iter=10000,
                                                      penalty='l1'),
                              param_grid=[{'alpha': [1e-06, 0.22222311111111112,
                                                     0.4444452222222222,
                                                     0.6666673333333334,
                                                     0.8888894444444445,
                                                     1.1111115555555555,
                                                     1.3333336666666666,
                                                     1.5555557777777778,
                                                     1.7777778888888889,
                                                     2.0]}],
                              scoring='f1_macro'))])

In [ ]:
best_param = clf_sgd_grid_search.named_steps['gridsearchcv'].best_params_
display(best_param)

{'alpha': 1e-06}

In [ ]:
clf_sgd_ = SGDClassifier(max_iter=10000, penalty='l1', loss='hinge', tol=1e-3, alpha=1e-06)
scores_sgd_ = cross_val_score(clf_sgd_, train_data, train_target, cv=10, scoring='f1_macro')
print("F1_Macro SGD", scores_sgd_.mean()," (+/-) ", scores_sgd_.std() * 2)

F1_Macro SGD 0.8079975075429621  (+/-)  0.09424587439343432


## Keras

In [ ]:
print("Keras")

Keras
